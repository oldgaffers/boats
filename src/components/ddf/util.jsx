export const mapPicker = (m) => {
  return m?.map((i) => {
    if (i.name) {
      return { label: i.name, value: i.id }
    }
    return { label: i.replace('_', ' '), value: i }
  }) || [];
}

export function intField(name, label) {
  return {
    name, label,
    component: "text-field", type: "number", dataType: 'integer',
  };
}

export const basicDimensionItems = [
  {
    component: 'text-field',
    name: "handicap_data.length_on_deck",
    label: "Length on deck (LOD) (decimal feet)",
    type: "number",
    dataType: 'float',
    isRequired: true,
    validate: [
      {
        type: 'required',
      },
    ],
    resolveProps: (props, { meta, input }, formOptions) => {
      const { values } = formOptions.getState();
      if (values.handicap_data?.length_on_deck === undefined) {
        return {
          initialValue: 30,
        };
      }
    },
  },
  {
    component: 'text-field',
    name: "handicap_data.length_on_waterline",
    label: "Waterline Length {LWL) (decimal feet)",
    type: "number",
    dataType: 'float',
    isRequired: true,
    validate: [
      {
        type: 'required',
      },
    ],
    resolveProps: (props, { meta, input }, formOptions) => {
      const { values } = formOptions.getState();
      if (values.handicap_data?.length_on_waterline === undefined) {
        return {
          initialValue: 28,
        };
      }
    },
  },
  {
    component: 'text-field',
    name: "handicap_data.beam",
    label: "Beam (decimal feet)",
    type: "number",
    dataType: 'float',
    isRequired: true,
    validate: [
      {
        type: 'required',
      },
    ],
    resolveProps: (props, { meta, input }, formOptions) => {
      const { values } = formOptions.getState();
      if (values.handicap_data?.beam === undefined) {
        return {
          initialValue: 8,
        };
      }
    },
  },
  {
    component: 'text-field',
    name: "handicap_data.draft",
    label: "Minumum Draft (decimal feet)",
    type: "number",
    dataType: 'float',
    isRequired: true,
    validate: [
      {
        type: 'required',
      },
    ],
    resolveProps: (props, { meta, input }, formOptions) => {
      const { values } = formOptions.getState();
      if (values.handicap_data?.draft === undefined) {
        return {
          initialValue: 4.5,
        };
      }
    },
  },
  {
    component: 'text-field',
    name: "air_draft",
    label: "Air Draft (decimal feet)",
    type: "number",
    dataType: 'float',
    validate: [
    ],
  },
  {
    component: 'plain-text',
    name: "ddf.h1",
    label: "LOD, LWL, beam and draft affect handicaps",
  },
];

export const constructionItems = (pickers) => {
  return [
    {
      component: 'select',
      name: "construction_material",
      label: "Construction material",
      isReadOnly: false,
      isSearchable: true,
      isClearable: true,
      options: mapPicker(pickers.construction_material),
      isOptionEqualToValue: (option, value) => option.value === value,
    },
    {
      component: 'select',
      name: "construction_method",
      label: "Construction method",
      isReadOnly: false,
      isSearchable: true,
      isClearable: true,
      options: mapPicker(pickers.construction_method),
      isOptionEqualToValue: (option, value) => option.value === value,
    },
    {
      component: 'select',
      name: "spar_material",
      label: "Spar material",
      isReadOnly: false,
      isSearchable: true,
      isClearable: true,
      options: mapPicker(pickers.spar_material),
      isOptionEqualToValue: (option, value) => option.value === value,
    },
    {
      component: 'text-field',
      name: "construction_details",
      label: "Construction details",
    },
  ];
};

function optionlist(p) {
  const l = [...new Set(p.filter((o) => o && o.trim() !== ''))];
  return l.map((o) => ({ label: o, value: o }));
}

function optionsFromPicker(p) {
  if (Array.isArray(p)) {
    if (p.length === 0) {
      return [];
    }
    if (typeof p[0] === 'string') {
      return optionlist(p);
    }
    return optionlist(p.map((o) => o.name));
  }
  return [];
}

export const extendableList = ({ pickers, name, label }) => {
  return [
    {
      component: 'dual-list-select',
      name,
      label: `${label}(s)`,
      options: optionsFromPicker(pickers[name]),
    },
    {
      component: 'text-field',
      name: `ddf.new_${name}`,
      label: `if the ${label.toLowerCase()} is not listed and you know the name add it here`,
      isRequired: false,
    },
  ];
};

export const extendableItems = ({ pickers, name, label }) => {
  // console.log('extendableItems', name, label)
  return [
    {
      component: 'select',
      name,
      label,
      isReadOnly: false,
      isSearchable: true,
      isClearable: true,
      options: pickers[name].map((o) => ({ label: o.name, value: o.name })),
      noOptionsMessage: 'we don\'t have that one - you can add it below',
    },
    {
      component: 'text-field',
      condition: {
        or: [
          { when: name, isEmpty: true },
          { when: name, is: ' ' }
        ]
      },
      name: `ddf.new_${name}`,
      label: `if you can't find anything suitable add a new one here`,
      isRequired: false,
    },
  ];
};

export const GenericTypeItems = (pickers) => extendableList({ pickers, name: 'generic_type', label: 'Generic Type' })
export const builderItems = (pickers) => extendableList({ pickers, name: 'builder', label: 'Builder' })
export const designerItems = (pickers) => extendableList({ pickers, name: 'designer', label: 'Designer' })
export const designClassItems = (pickers) => extendableItems({ pickers, name: 'design_class', label: 'Design Class' })
