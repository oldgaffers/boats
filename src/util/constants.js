export const boatRegisterHome = "https://oldgaffers.github.io";

