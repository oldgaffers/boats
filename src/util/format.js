const OneFoot = 0.3048;

const pounds = new Intl.NumberFormat('en-GB', { currency: 'GBP', style: 'currency' });

export function formatList(b, k) {
  const data = b?.[k];
  if (data?.name) {
    return data.name;
  }
  if (Array.isArray(data)) {
    return data.map((d) => d?.name || d).join(' / ');
  }
}

export function price(n) {
  if (n === 0) {
    return 'offers';
  }
  return pounds.format(n);
}

export function m2dfn(val) {
  if (val) {
    return Math.round(1000 * val / OneFoot) / 1000;
  }
}

export function m2df(val) {
  if (val) {
    //  return (val/OneFoot).toFixed(2);
    return m2dfn(val).toFixed(2);
  }
}

export function feet(n) {
  return `${n.toFixed(2)} ft`;
}

export function squarefeet(n) {
  return `${n.toFixed(2)} ftÂ²`;
}

export function kg(n) {
  if (n) {
    return `${Math.floor(n)} kg`;
  }
}

export function m2f(val) {
  if (val) {
    // return `${m2df(val)} ft`
    return feet(m2dfn(val))
  }
};

export function m2f2(val) {
  if (val) {
    return squarefeet(val / OneFoot / OneFoot);
  }
};

export function m2dsqf(val) {
  if (val) {
    return (val / OneFoot / OneFoot).toFixed(3);
  }
}

export function m2dsqfn(val) {
  if (val) {
    return parseFloat((val / OneFoot / OneFoot).toFixed(3));
  }
}

export function f2m(val) {
  if (val) {
    return Math.round(1000 * OneFoot * val) / 1000;
  }
}

export function f2m2(val) {
  if (val) {
    return Math.round(1000 * OneFoot * OneFoot * val) / 1000;
  }
}

const metreKeys = [
  'beam', 'draft', 'draft_keel_down', 'air_draft',
  'perpendicular', 'luff', 'head', 'leech', 'foot',
  'length_on_deck', 'length_on_waterline', 'length_over_all',
  'length_over_spars',
  'fore_triangle_height', 'fore_triangle_base',
];
const squareMetreKeys = ['sailarea'];
const booleanKeys = ['year_is_approximate'];

export function boatm2f(obj) {
  if (obj) {
    if (Array.isArray(obj)) {
      return obj.map((n) => boatm2f(n))
    } else if (typeof obj === 'object') {
      const r = {};
      Object.keys(obj).forEach(k => {
        if (metreKeys.includes(k)) {
          r[k] = m2dfn(obj[k]);
        } else if (squareMetreKeys.includes(k)) {
          r[k] = m2dsqfn(obj[k]);
        } else if (booleanKeys.includes(k)) {
          r[k] = !!obj[k];
        } else if (obj[k]) {
          r[k] = boatm2f(obj[k]);
        }
      });
      return r;
    } else {
      return obj;
    }
  } else {
    return undefined;
  }
}

export function boatf2m(obj) {
  if (obj) {
    if (Array.isArray(obj)) {
      return obj.map((n) => boatm2f(n))
    } else if (typeof obj === 'object') {
      const r = {};
      Object.keys(obj).forEach(k => {
        if (metreKeys.includes(k)) {
          r[k] = f2m(obj[k]);
        } else if (squareMetreKeys.includes(k)) {
          r[k] = f2m2(obj[k]);
        } else if (booleanKeys.includes(k)) {
          r[k] = !!obj[k];
        } else {
          if (obj[k]) {
            r[k] = boatf2m(obj[k]);
          } else {
            r[k] = obj[k];
          }
        }
      });
      return r;
    } else {
      return obj;
    }
  }
  return obj;
}

export function boatDefined(value) {
  if (value === false) {
    return false;
  }
  if (!value) {
    return undefined;
  }
  if (typeof value === 'object') {
    if (Array.isArray(value)) {
      if (value.length === 0) {
        return undefined;
      }
      return value.map((n) => boatDefined(n)).filter((item) => item);
    }
    if (Object.keys(value).length > 0) {
      return Object.fromEntries(
        Object.keys(value)
          .map((key) => [key, boatDefined(value[key])])
          .filter(([key, value]) => value !== undefined)
      );
    }
  }
  return value;
}

export function m2fall(o) {
  if (o) {
    return Object.keys(o).map(k => m2dfn(o[k]));
  }
}

export function newestForSaleRecord(boat) {
  return boat?.for_sales?.reduce((prev, curr) =>
    (new Date(prev.created_at)
      >
      new Date(curr.created_at)
    ) ? prev : curr,
    { created_at: '1970-01-01T00:00:00Z' },
  );
}