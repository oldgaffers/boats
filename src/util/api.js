import boatsByPlaceBuilt from './boatsByPlaceBuilt';
import boatsByHomePort from './boatsbyhomeport';
import { boatRegisterHome } from './constants';

const mock = false; // true;

const mocks = {
  filterable: [
    {
      oga_no: 315,
      name: 'Robinetta',
    },
  ],
  boat: {
    oga_no: 315,
    name: 'Robinetta',
  },
};

const api1 = 'https://3q9wa2j7s1.execute-api.eu-west-1.amazonaws.com';
const api2 = 'https://v2z7n3g4mb5lfkbgeyln3ksjya0qsrpm.lambda-url.eu-west-1.on.aws';
const stage = 'default'

export async function putGeneralEnquiry(scope, subject, data) {
  return (await fetch(
    `${api1}/${stage}/${scope}/${subject}`,
    {
      method: 'PUT',
      body: JSON.stringify(data),
      headers: { 'content-type': 'application/json' }
    }
  )).json();
}

export async function shuffleBoats() {
  return (await fetch(
    `${api1}/${stage}/shuffle`,
    {
      method: 'POST',
    }
  )).json();
}

export async function postBoatData(data) {
  // console.log('postBoatData', data);
  // return { ok: true };
  return fetch(
    'https://t5lcpwbufg6ewyxswawwsbtkjq0kupbb.lambda-url.eu-west-1.on.aws/',
    {
      method: 'POST',
      body: JSON.stringify(data),
      headers: { 'content-type': 'application/json' }
    }
  );
}

export async function postScopedData(scope, subject, data, accessToken) {
  const headers = {
    'content-type': 'application/json',
  };
  if (accessToken) {
    headers.Authorization = `Bearer ${accessToken}`;
  }
  return fetch(
    `${api1}/${stage}/${scope}/${subject}`,
    {
      method: 'POST',
      body: JSON.stringify(data),
      headers,
    },
  );
}

export async function postGeneralEnquiry(scope, subject, data) {
  return fetch(
    `${api1}/${stage}/${scope}/${subject}`,
    {
      method: 'POST',
      body: JSON.stringify(data),
      headers: { 'content-type': 'application/json' }
    }
  );
}

export async function postCrewEnquiry(data) {
  return fetch(
    'https://bne25yrc5scp5qgffis2yfqaj40txzef.lambda-url.eu-west-1.on.aws/',
    {
      method: 'POST',
      body: JSON.stringify(data),
      headers: { 'content-type': 'application/json' }
    }
  );
}

export async function getBoatData(ogaNo) {
  if (mock) {
    return mocks.boat;
  }
  const r = await fetch(`${boatRegisterHome}/boatregister/page-data/boat/${ogaNo}/page-data.json`);
  if (r.ok) {
    const data = await r.json();
    data.result.pageContext.boat?.for_sales?.sort((a, b) => Date.parse(b.created_at) - Date.parse(a.created_at));
    ['generic_type', 'builder', 'designer'].forEach((key) => {
      if (!Array.isArray(data.result.pageContext.boat[key])) {
        data.result.pageContext.boat[key] = [data.result.pageContext.boat[key]];
      }
      data.result.pageContext.boat[key] = data.result.pageContext.boat[key].filter((v) => v);
    });
    return data;  
  }
}

export async function getPicklists() {
  if (mock) {
    return {};
  }
  return (await fetch(`${boatRegisterHome}/boatregister/pickers.json`)).json();
}

export async function getExtra() {
  /* try {
    return getScopedData('public', 'crewing');
  } catch(e) {
    console.log(e);
  }
    */
  return undefined;
}

export async function getFilterable() {
  if (mock) {
    return mocks.filterable;
  }
  const extra = await getExtra();
  // console.log('crewing', extra);
  const ex = Object.fromEntries((extra?.Items || []).map((item) => [item.oga_no, item]));
  const filterable = await (await fetch(`${boatRegisterHome}/boatregister/filterable.json`)).json();
  return filterable.map((b) => {
    if (ex[b.oga_no]) {
      return { ...b, ...ex[b.oga_no]};
    }
    return b;
  });
}

export async function getPlaces() {
  return (await fetch(`${boatRegisterHome}/boatregister/places.json`)).json();
}

export async function getUploadCredentials() {
  return (await fetch(`${api1}/${stage}/upload_credentials`)).json();
}

export async function createPhotoAlbum(name, ogaNo) {
  return fetch(`${api2}/`,
    {
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ name, oga_no: ogaNo }),
      method: 'POST',
    }
  );
}

export async function getAlbumKey(name, oga_no) {
  const r = await fetch(`${api2}/album?oga_no=${oga_no}&name=${encodeURIComponent(name)}`);
  if (r.ok) {
    return r.json();
  }
  return undefined;
}

export async function getThumb(albumKey) {
  return (await fetch(`${api2}/thumb?album_key=${albumKey}`)).json();
}

export async function getLargestImage(albumKey) {
  return (await fetch(`${api2}/li?album_key=${albumKey}`)).json();
}

export async function nextOgaNo() {
  const url = 'https://fxaj7udnm64v43j6fjo4zqer5u0xmhra.lambda-url.eu-west-1.on.aws';
  return (await fetch(`${url}/`)).json();
}

export async function disposeOgaNo(oga_no) {
  const r = await fetch(
    'https://fxaj7udnm64v43j6fjo4zqer5u0xmhra.lambda-url.eu-west-1.on.aws/',
    {
      method: 'POST',
      body: `${oga_no}`,
      headers: { 'content-type': 'application/json' }
    }
  );
  return r.json();
}

const localTables = [
  { subject: 'place_built', action: boatsByPlaceBuilt },
  { subject: 'home_port', action: boatsByHomePort },
];

export async function getScopedData(scope, subject, filters, accessToken) {
  if (mock) {
    return {};
  }
  const local = localTables.find((l) => l.subject === subject);
  if (local) {
    return local.action(filters);
  }
  if (subject === 'place_built') {
    return boatsByPlaceBuilt(filters);
  }
  const headers = {
    'content-type': 'application/json',
  };
  if (accessToken) {
    headers.Authorization = `Bearer ${accessToken}`;
  }
  const r = await fetch(
    `${api1}/${stage}/${scope}/${subject}?${new URLSearchParams(filters)}`,
    {
      headers,
    });
  return r.json();
}

export async function getFleets(scope, filter, accessToken) {
  const r = await getScopedData(scope, 'fleets', filter, accessToken);
  const f = r?.Items || [];
  f.sort((a,b) => a.name.localeCompare(b.name))
  return f;
}

export async function getBoatLastModified(oga_no) {
  const r = await fetch(`https://api.github.com/repos/oldgaffers/boatregister/commits?path=boat/${oga_no}/boat.yml&page=1&per_page=1`);
  const d = await r.json();
  return d[0].commit.author.date;
}

export async function geolocate(place) {
  const r = await fetch(`${api1}/${stage}/public/place?name=${place}`);
  if (r.ok) {
    return r.json()
  }
  return undefined;
}