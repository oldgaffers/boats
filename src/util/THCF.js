import { f2m } from "./format";
// Constants
const baseLengthInFeetForTHCF = 25;

// Helper Functions
function rigAllowance(r) {
  r = r.toLowerCase();
  switch (r) {
    case 'cutter': return 0.96;
    case 'yawl': return 0.94;
    case 'schooner': return 0.92;
    case 'ketch': return 0.90;
    default: return 1.0;
  }
}

export function sails(r) {
  r = r.toLowerCase();
  if (['cutter', 'sloop'].includes(r)) {
    return ['fore_triangle', 'main', 'topsail'];
  } else if (r === 'yawl' || r === 'ketch') {
    return ['fore_triangle', 'main', 'topsail', 'mizzen', 'mizzen_topsail'];
  } else if (r === 'schooner') {
    return ['fore_triangle', 'main', 'topsail', 'fore_sail', 'fore_topsail', 'mizzen', 'mizzen_topsail'];
  } else if (['catboat', 'single sail'].includes(r)) {
    return ['main'];
  }
  return [];
}

export function sailArea(boat) {
  const s = sails(boat.rig_type || '');
  const sailAreas = {};
  const hd = boat.handicap_data || {};
  if (!hd.main) return hd.sailarea || 0;

  for (const sail of s) {
    sailAreas[sail] = 0;
    if (sail === 'fore_triangle') {
      sailAreas[sail] = fForeTriangle(hd);
    } else if (['main', 'fore_sail', 'mizzen'].includes(sail)) {
      sailAreas[sail] = fMainSA(hd[sail] || {});
    } else if (['topsail', 'fore_topsail', 'mizzen_topsail'].includes(sail)) {
      sailAreas[sail] = fTopSA(hd[sail] || {});
    }
  }

  const total = Object.values(sailAreas).reduce((a, b) => a + b, 0);
  return total || 0;
}

export function fMainSA(sail) {
  if (!sail) return 0;
  const b = sail.foot || 0;
  const h = sail.luff || 0;
  if (b == null || h == null) return 0;
  if (!('head' in sail)) return 0.5 * b * h;
  const g = parseFloat(sail.head);
  const d = Math.sqrt(b * b + h * h);
  return 0.5 * b * h + 0.5 * g * d;
}

export function fTopSA(sail) {
  if (sail) {
    const i = sail.perpendicular || 0;
    const h = sail.luff || 0;
    if (i != null && h != null) return 0.5 * i * h;
  }
  return 0;
}

export function fForeTriangle(data) {
  if (data) {
    const i = parseFloat(data.fore_triangle_height || 0);
    const j = parseFloat(data.fore_triangle_base || 0);
    return 0.425 * i * j;
  }
  return 0;
}

export function fL(data) {
  if (data?.length_on_deck && data?.length_on_waterline) {
    return 0.5 * (data.length_on_deck + data.length_on_waterline);
  }
  return 0;
}

export function fBD(boat) {
  const hd = boat?.handicap_data;
  if (hd?.beam) return 0.67 * hd.beam * hd.beam;
  return 0;
}

export function fSqrtS(rigAllowanceVal, sailarea) {
  return rigAllowanceVal * Math.sqrt(sailarea);
}

export function fMR(boat, useSailArea = false) {
  const hd = boat?.handicap_data;
  if (!boat.rig_type || !hd) return 0;
  const L = fL(hd);
  let sailAreaVal = sailArea(boat);
  if (sailAreaVal === 0 && useSailArea && hd.sailarea) sailAreaVal = hd.sailarea;
  if (sailAreaVal === 0) return 0;
  const sqrtS = fSqrtS(rigAllowance(boat.rig_type), sailAreaVal);
  if (sqrtS === 0 || L === 0) return 0;
  const BD = fBD(boat);
  if (BD > 0) {
    const x = 0.15 * L * sqrtS / Math.sqrt(BD);
    const y = 0.2 * (L + sqrtS);
    return x + y;
  }
  return 0;
}

export function fPropellorBonus(data) {
  const propType = data?.propellor?.type || '';
  if (propType === 'fixed') return 0.03;
  if (['folding', 'feathering'].includes(propType)) return 0.015;
  return 0;
}

export function fShoalBonus(R, boat) {
  const hd = boat?.handicap_data;
  if (!hd) return 0;
  const lwl = hd.length_on_waterline;
  const draft = boat.draft;
  const nominalDraft = 0.12 * lwl;
  let bonus = 0;
  if (draft < 0.8 * nominalDraft) bonus = 0.02 * R;
  if (draft < 0.66 * nominalDraft) bonus = 0.02 * R;
  if (['centre-boarder', 'centreboard dinghy', 'leeboarder'].includes(boat.hull_form)) bonus /= 2;
  return bonus;
}

export function fR(boat, useSailArea = false ) {
  if (!boat) return 0;
  const MR = fMR(boat, useSailArea);
  return MR - MR * fPropellorBonus(boat.handicap_data);
}

export function fThcf(boat, useSailArea = false) {
  const r = fR(boat, useSailArea);
  if (r <= 0) return 0;
  return 0.125 * (Math.sqrt(r) + 3);
}
              
/*
Proposed rating formula (metres):
Measured Rating = MR = 0.2*(L*√S)/√(B*D*SF) + 0.67*(L+√S)
OR if displacement is known:
Measured Rating = MR = 0.2*(L*√S)/√(Disp/L) + 0.67*(L+√S)
Where:
L = (LWL+LOD)/2
S = Corrected sail area (includes rig allowance)
B = Beam (Should be water line, but in most cases the difference is insignificant)
D = Draft
SF = Shape Factor above
Disp = Displacement

*/
const shapeFactorMap = {
    'Long keel - High volume': 0.25,
    'Long keel - Standard': 0.20,
    'Long keel - Low volume': 0.15,
    'Fin keel': 0.10,
  };

export function shapeFactors(sf) {
    return shapeFactorMap?.[sf] || 0.2;
}

export function solentLength(data) {
    const LOD = data.length_on_deck || baseLengthInFeetForTHCF;
    const LWL = data.length_on_waterline || baseLengthInFeetForTHCF;
    return f2m(0.5*(LOD+LWL));
}

// normalised to kg using 1000 kg per cubic metre for salt water
export function solentEstimatedDisplacement(data) {
    const L = solentLength(data);
    const B = f2m(data.beam);
    const D = f2m(data.draft);
    const SF = shapeFactors(data.solent.hull_shape);
    return Math.round(1000 * L * B * D * SF);
}

export function solentMR(boat) {
    const { ddf, handicap_data } = boat;
    const L = solentLength(handicap_data);
    const rS = f2m(ddf.root_s || 0.0);
    const y = 0.67 * (L + rS);
    if (handicap_data.displacement) {
        const enteredDisplacement = handicap_data.displacement / 1000; // in cubic metres
        const x = 0.2 * L * rS / Math.sqrt(enteredDisplacement / L);
        return x + y;    
    } else {
        // const estimatedDisplacement = solentEstimatedDisplacement(data) / 1000; // in cubic metres
        // const x2 = 0.2 * L * rS / Math.sqrt(estimatedDisplacement / L);
        const B = f2m(handicap_data.beam);
        const D = f2m(handicap_data.draft);
        const SF = shapeFactors(handicap_data.solent.hull_shape);   
        const x = 0.2 * L * rS / Math.sqrt(B*D*SF);
        return x + y;            
    }
}

export function solentRating(boat) {
    const data = boat.handicap_data;
    const mmrf = data.solent.measured_rating;
    const thcf = fThcf(mmrf*(1 - boat.ddf.prop_allowance));
    const pf = Number(data.solent?.performance_factor || 0);
    const sr = (1 + pf) * thcf;
    return Math.round(1000 * sr) / 1000;
}