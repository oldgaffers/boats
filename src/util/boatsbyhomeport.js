import { geolocate, getFilterable } from "./api";

export default async function boatsByHomePort() {
    const boats = await getFilterable()
    const filtered = boats.filter((b) =>(b.home_port||'') !== '');
    const located = await Promise.all(filtered.map(async (b) => {
        const g = await geolocate(b.home_port);
        if (g?.lat) {
            return { ...b, ...g, latitude: g.lat, longitude: g.lng };
        }
        return null;
    }));
    const valid = (await located).filter((p) => p !== null);
    return { Items: valid };
}