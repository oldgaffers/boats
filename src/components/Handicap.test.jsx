import { foretriangle_area, mainsail_area } from './Handicap';
import { fMainSA, fTopSA, fForeTriangle, sailArea, fL, fMR, fBD, fSqrtS, fR, fThcf } from '../util/THCF';
import { boatm2f, boatf2m } from "../util/format";

const Transcur = {
  "builder": {
    "id": "78bc9e97-d57d-481e-b855-b10261a92e4d",
    "name": "Aldous Ltd."
  },
  "construction_details": "Trad ",
  "construction_material": "wood",
  "construction_method": "carvel",
  "created_at": "2020-04-25T17:32:47.429243+00:00",
  "designer": {
    "id": "74db1b7f-49fa-49d4-8e3f-9f172cd58673",
    "name": "Aldous"
  },
  "for_sales": [],
  "full_description": 'x',
  "generic_type": ["Smack"],
  "handicap_data": {
    "beam": 3.28,
    "draft": 1.396,
    "fore_triangle_base": 7.315,
    "fore_triangle_height": 10,
    "length_on_deck": 10.796,
    "length_on_waterline": 8.687,
    "length_over_all": 10.79,
    "main": {
      "foot": 7,
      "head": 5,
      "luff": 5.486
    },
    "propellor": {
      "type": "folding"
    },
    "sailarea": 81,
    "thcf": 1.025,
    "topsail": {
      "luff": 6.706,
      "perpendicular": 2.438
    }
  },
  "home_country": "GBR",
  "home_port": "Pin Mill",
  "hull_form": "long keel deep forefoot",
  "id": "656e0e94-c298-408e-90e2-1bafd5bdd824",
  "image_key": "grRdbV",
  "mainsail_type": "gaff",
  "name": "Transcur",
  "oga_no": 2932,
  "ownerships": [
    {
      "name": "Frank Mulville"
    },
    {
      "name": "Peter Maynard"
    },
    {
      "name": "Rober Bettle"
    },
    {
      "current": true,
      "id": 38,
      "member": 954,
      "share": 64,
      "start": 1998
    }
  ],
  "place_built": "Brightlingsea",
  "rig_type": "Cutter",
  "sail_number": "CK365",
  "selling_status": "not_for_sale",
  "short_description": 'x',
  "spar_material": "wood",
  "updated_at": "2022-10-30T06:56:03.964565+00:00",
  "year": 1889,
  "year_is_approximate": false
};

const Robinetta = {
  "builder": {
    "id": "a27bca50-d155-428a-be49-a542d64ba316",
    "name": "Enterprise Small Craft"
  },
  "callsign": "MKDZ8",
  "construction_details": "pitch pine on oak and elm",
  "construction_material": "wood",
  "construction_method": "carvel",
  "created_at": "2020-04-25T16:47:00.543376+00:00",
  "designer": {
    "id": "afc68d59-e7ba-466e-861f-741e2eeae526",
    "name": "D.A. Rayner"
  },
  "for_sales": [],
  "full_description": 'x',
  "generic_type": ["Yacht"],
  "handicap_data": {
    "beam": 2.441,
    "draft": 1.372,
    "fore_triangle_base": 1.524,
    "fore_triangle_height": 4.572,
    "length_on_deck": 6.858,
    "length_on_waterline": 6.096,
    "length_over_all": 6.858,
    "main": {
      "foot": 4,
      "head": 4.267,
      "luff": 4.319
    },
    "propellor": {
      "type": "fixed"
    },
    "sailarea": 24.156,
    "thcf": 0.85
  },
  "home_country": "GBR",
  "home_port": "Rosneath",
  "hull_form": "long keel deep forefoot",
  "id": "85315666-4359-47f7-8d8d-4dd2905b7aa2",
  "image_key": "9NV8KW",
  "mainsail_type": "gaff",
  "mmsi": "235109366",
  "name": "Robinetta",
  "nsbr": "1717",
  "oga_no": 315,
  "ownerships": [
    {
      "end": 1946,
      "name": "Denys Rayner",
      "share": 64,
      "start": 1937
    },
    {
      "end": 1949,
      "name": "Audrey Parker",
      "share": 32,
      "start": 1946
    },
    {
      "end": 1949,
      "name": "Henry Parker",
      "share": 32,
      "start": 1946
    },
    {
      "end": 1985,
      "name": "Nigel Heriot",
      "share": 64,
      "start": 1949
    },
    {
      "end": 1986,
      "name": "Frank Driscoll",
      "share": 64,
      "start": 1985
    },
    {
      "end": 2007,
      "name": "Mike Garnham",
      "share": 64,
      "start": 1986
    },
    {
      "current": true,
      "id": 559,
      "member": 5004,
      "share": 32,
      "start": 2007
    },
    {
      "current": true,
      "id": 1219,
      "member": 5004,
      "share": 32,
      "start": 2007
    }
  ],
  "place_built": "Rock Ferry, Birkenhead",
  "reference": [
    "YM May 1937",
    "YM Nov 1937",
    "YM Nov 1947",
    "YM Nov 2022"
  ],
  "rig_type": "Cutter",
  "sail_number": "315",
  "selling_status": "not_for_sale",
  "short_description": 'x',
  "spar_material": "wood",
  "ssr": "131134",
  "uk_part1": "165315",
  "updated_at": "2022-10-30T08:21:32.430046+00:00",
  "website": "https://robinetta-log.blogspot.com/p/robinetta.html",
  "year": 1937,
  "year_is_approximate": false
};

const tcf = boatm2f(Transcur).handicap_data;
const rcf = boatm2f(Robinetta).handicap_data;

test('boatm2f', () => {
  const a = { "1": 1, q: { a: 2 } };
  expect(boatm2f(a)).toStrictEqual(a);
});

test('boatm2f Transcur feet', () => {
  expect(tcf).toMatchSnapshot();
});

test('boatm2f Robinetta feet', () => {
  expect(rcf).toMatchSnapshot();
});

test('boatm2f Transcur metres', () => {
  expect(boatf2m(boatm2f(Transcur))).toStrictEqual(Transcur);
});

test('boatm2f Robinetta metres', () => {
  expect(boatf2m(boatm2f(Robinetta))).toStrictEqual(Robinetta);
});

test('Transcur foretriangle_area', () => {
  const f = foretriangle_area(tcf);
  expect(f.toFixed(1)).toMatchSnapshot();
});

test('Robinetta foretriangle_area', () => {
  const f = foretriangle_area(rcf);
  expect(f).toMatchSnapshot();
});

test('Transcur fMainSA', () => {
  expect(fMainSA(tcf.main).toFixed(2)).toMatchSnapshot();
});
test('Transcur fTopSA', () => {
  expect(fTopSA(tcf.topsail).toFixed()).toMatchSnapshot();
});

test('Transcur fForeTriangle', () => {
  expect(fForeTriangle(tcf).toFixed(2)).toMatchSnapshot();
});

test('Transcur fMSA', () => {
  expect(sailArea(Transcur).toFixed(1)).toMatchSnapshot();
});

test('Transcur fL', () => {
  expect(fL(tcf)).toEqual((tcf.length_on_waterline + tcf.length_on_deck) / 2);
});

test('Transcur fBD', () => {
  const b = boatm2f(Transcur);
  expect(fBD(b)).toMatchSnapshot();
});

test('Transcur fMR', () => {
  const b = boatm2f(Transcur);
  expect(fMR(b).toFixed(2)).toMatchSnapshot();
});

test('Transcur T(H)CF', () => {
  const b = boatm2f(Transcur);
  expect(fThcf(b).toFixed(3)).toMatchSnapshot();
});

test('Robinetta T(H)CF', () => {
  const b = boatm2f(Robinetta);
  expect(fThcf(b).toFixed(3)).toMatchSnapshot();
});
