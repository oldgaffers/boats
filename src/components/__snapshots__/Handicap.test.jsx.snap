// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Robinetta T(H)CF 1`] = `"0.850"`;

exports[`Robinetta foretriangle_area 1`] = `2.961`;

exports[`Transcur C 1`] = `"NaN"`;

exports[`Transcur T(H)CF 1`] = `"1.012"`;

exports[`Transcur fForeTriangle 1`] = `"31.09"`;

exports[`Transcur fGaffSA 1`] = `"41.44"`;

exports[`Transcur fMR 1`] = `"26.36"`;

exports[`Transcur fMSA 1`] = `"72.5"`;

exports[`Transcur fR 1`] = `"25.96"`;

exports[`Transcur fSqrtS 1`] = `"26.82"`;

exports[`Transcur fTopSA 1`] = `"8"`;

exports[`Transcur foretriangle_area 1`] = `"31.1"`;

exports[`boatm2f Robinetta feet 1`] = `
Object {
  "beam": 8.009,
  "draft": 4.501,
  "fore_triangle_base": 5,
  "fore_triangle_height": 15,
  "length_on_deck": 22.5,
  "length_on_waterline": 20,
  "length_over_all": 22.5,
  "main": Object {
    "foot": 13.123,
    "head": 13.999,
    "luff": 14.17,
  },
  "propellor": Object {
    "type": "fixed",
  },
  "sailarea": 260.013,
  "thcf": 0.85,
}
`;

exports[`boatm2f Transcur feet 1`] = `
Object {
  "beam": 10.761,
  "draft": 4.58,
  "fore_triangle_base": 23.999,
  "fore_triangle_height": 32.808,
  "length_on_deck": 35.42,
  "length_on_waterline": 28.501,
  "length_over_all": 35.4,
  "main": Object {
    "foot": 22.966,
    "head": 16.404,
    "luff": 17.999,
  },
  "propellor": Object {
    "type": "folding",
  },
  "sailarea": 871.877,
  "thcf": 1.025,
  "topsail": Object {
    "luff": 22.001,
    "perpendicular": 7.999,
  },
}
`;

exports[`t1 1`] = `1.025`;
